nebula-logger/main/log-management/classes/LogEntryEventHandler.cls

```diff
@@ -218,64 +221,173 @@ public without sharing class LogEntryEventHandler extends LoggerSObjectHandler {
-        for (LogEntry__c logEntry : this.logEntryToTopics.keySet()) {
-            if (this.logEntryToTopics.containsKey(logEntry) == false) {
-                continue;
+        if (!missingTagsToCreate.isEmpty()) {
+            insert missingTagsToCreate;
+            for (SObject tag : missingTagsToCreate) {
+                missingTagNameToId.put((String) tag.get('Name'), (Id) tag.get('Id'));
             }
+        }
+
+        return missingTagNameToId;
+    }
+
+    // Private static methods
+    private static String getLogEntryKey(LogEntry__c logEntry) {
+        // TODO - after upgrading all metadata to v52.0, switch to using the new EventUuid field for platform events
+        // https://releasenotes.docs.salesforce.com/en-us/summer21/release-notes/rn_batch_platform_event.htm
+        return logEntry.Log__r.TransactionId__c + logEntry.TransactionEntryNumber__c;
+    }
 
-            for (String topicName : this.logEntryToTopics.get(logEntry)) {
-                // Add log entry-specific topics
-                topicAssignments.add(new TopicAssignment(EntityId = logEntry.Id, TopicId = topicNameToTopics.get(topicName).Id));
+    private static List<String> getTagNames(String tagsString) {
+        List<String> cleanedTagNames = new List<String>();
 
-                // Add all topics to the parent log when enabled
-                topicAssignments.add(new TopicAssignment(EntityId = logEntry.Log__c, TopicId = topicNameToTopics.get(topicName).Id));
+        for (String tagName : tagsString.split('\n')) {
+            if (String.isNotBlank(tagName) == true) {
+                cleanedTagNames.add(tagName.trim());
             }
         }
-        insert new List<TopicAssignment>(topicAssignments);
+
+        return cleanedTagNames;
+    }
+
+    private static List<LogEntryTagRule__mdt> getTagAssignmentRules() {
+        List<LogEntryTagRule__mdt> tagAssignmentRules = [
+            SELECT Id, SObjectField__r.QualifiedApiName, ComparisonType__c, ComparisonValue__c, Tags__c
+            FROM LogEntryTagRule__mdt
+            WHERE IsEnabled__c = TRUE AND SObjectType__r.DeveloperName = 'LogEntry'
+        ];
+        for (LogEntryTagRule__mdt rule : tagAssignmentRules) {
+            rule.SObjectField__c = rule.SObjectField__r.QualifiedApiName;
+        }
+        if (Test.isRunningTest() == true) {
+            // During tests, only use mock records - tests can add mock records using LogEntryEventHandler.TAG_ASSIGNMENT_RULES.add()
+            tagAssignmentRules.clear();
+        }
+        return tagAssignmentRules;
+    }
+
+    private static Boolean ruleCriteriaMet(LogEntry__c logEntry, LogEntryTagRule__mdt rule) {
+        Boolean ruleCriteriaMet = false;
+        String logEntryFieldValue = String.valueOf(logEntry.get(rule.SObjectField__c));
+        String ruleComparisonValue = rule.ComparisonValue__c;
+
+        switch on rule?.ComparisonType__c.toUpperCase() {
```

Comment:
is there a test for comparison type being null? I haven't tried switching on null when the type wasn't known ...

Reply:
There's not a test at the moment - it should just get ignored (since it doesn't match one of the specified Strings in the switch statement), but I'll add one just to be sure that it doesn't break anything

Reply:
Actually, I'm going to skip adding a test for null - the field `ComparisonType__c` is required, so null wouldn't actually happen.

Nevertheless, I wrote this small `switch` statement to show how nulls are handled - the null falls into the `else` block. If there weren't an `else` section, then the null value would just be ignored altogether by the `switch` statement
```java
String someString = null;
switch on someString {
    when 'something' {
        System.debug('someString == something');
    }
    when else {
        System.debug('here we are in the ELSE block, someString==' + someString);
    }
}
```

Reply:
Ah yeah, I was aware of the behavior of null in the when blocks - I guess I was just surprised that the compiler actually knows that safe navigation on a case statement that results in null is fine (basically it seems it must infer the type from the field); in other words, I don’t think that `switch on null` is currently valid, but clearly there was some thought put into how safe navigation works there